grakn-kb/src/main/java/ai/grakn/kb/internal/concept/ThingImpl.java

```diff
@@ -258,6 +260,35 @@ public T attribute(Attribute attribute){
         return getThis();
     }
 
+    @Override
+    public T deleteAttribute(Attribute attribute){
+        Role roleHasOwner = vertex().tx().getSchemaConcept(Schema.ImplicitType.HAS_OWNER.getLabel(attribute.type().getLabel()));
+        Role roleKeyOwner = vertex().tx().getSchemaConcept(Schema.ImplicitType.KEY_OWNER.getLabel(attribute.type().getLabel()));
+
+        Role roleHasValue = vertex().tx().getSchemaConcept(Schema.ImplicitType.HAS_VALUE.getLabel(attribute.type().getLabel()));
+        Role roleKeyValue = vertex().tx().getSchemaConcept(Schema.ImplicitType.KEY_VALUE.getLabel(attribute.type().getLabel()));
+
+        Stream<Relationship> relationships = validStreamOf(this::relationships, roleHasOwner, roleKeyOwner);
+        relationships.filter(relationship -> {
+                    Stream<Thing> rolePlayers = validStreamOf(relationship::rolePlayers, roleHasValue, roleKeyValue);
+                    return rolePlayers.anyMatch(rolePlayer -> rolePlayer.equals(attribute));
+                }).forEach(Concept::delete);
+
+        return getThis();
+    }
+
+    /**
+     * Returns a stream computed from a <code>streamProvider</code> which takes in as input an array or {@link Role}s
+     * which cannot contain any nulls.
+     *
+     * This helper exists to filter away the nulls
+     */
+    private <X> Stream<X> validStreamOf(Function<Role[], Stream<X>> streamProvider, Role ... roles){
+        if(roles.length == 0) return Stream.empty();
+        Role[] filteredRoles = Arrays.stream(roles).filter(Objects::nonNull).toArray(Role[]::new);
+        return streamProvider.apply(filteredRoles);
+    }
```

Comment:
I wouldn't bother passing the lambda in here, just make a method `Role[] filterNullls(Role... roles)`:
```
Stream<Relationship> relationships = relationships(filterNulls(roleHasOwner, roleKeyOwner));
```